### PR TITLE
[ADD] l10n_ro_edi: CIUS-RO export

### PR DESCRIPTION
**Pull request by yosuanicolaus**
Description of the issue/feature this PR addresses:
- In Romania, generating and communicating XML invoices to the government is a requirement. This format is based on UBL/CII for which we already have a very strong base.
- This task aims at completing the localisation with a first step : make Odoo generate the required Romanian UBL format if activated
- Official documentation: https://mfinante.gov.ro/web/efactura/informatii-tehnice

Current behavior before PR:
- CIUS_RO XML export is not supported yet

Desired behavior after PR is merged:
- CIUS_RO XML export will be supported

Task-id: 3388222

### CODE REVIEW

**Review by malb-odoo — CHANGES_REQUESTED** [review]
Good job ! Few small comment :smile: 
Suggestion: For this kind of pr where there is a official document, add the link in the commit message, 

Also don't forget to squash the different commit :smile:

File: addons/l10n_ro_edi/i18n/ro.po [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+"Language-Team: \n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=UTF-8\n"
+"Content-Transfer-Encoding: \n"
+"Plural-Forms: \n"
+
+#. module: l10n_ro_edi
+#: model:ir.model,name:l10n_ro_edi.model_account_edi_xml_ubl_ro
+msgid "CIUS RO"
+msgstr ""
```
Comment (malb-odoo) [inline addons/l10n_ro_edi/i18n/ro.po:21]: add Romanian translation (google translate is enough)

File: addons/l10n_ro_edi/tests/test_xml_ubl_ro.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+            'amount': 19,
+            'type_tax_use': 'sale',
+            'country_id': cls.env.ref('base.ro').id,
+        })
+
+    ####################################################
+    # Test export - import
+    ####################################################
+
+    def create_post_send_invoice(self, move_type):
```
Comment (malb-odoo) [inline addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:57]: I would rather say create_post_send_move since you don't use it only for invoices

File: addons/l10n_ro_edi/tests/test_xml_ubl_ro.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        invoice = self.env["account.move"].create({
```
Comment (malb-odoo) [inline addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:58]: So you could rename invoice by move or inline everything and remove the variable it's up to you :smile:

File: addons/l10n_ro_edi/tests/test_xml_ubl_ro.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+
+    def test_export_without_country_code_prefix_in_vat(self):
+        self.company_data['company'].vat = '1234567897'
+        invoice = self.create_post_send_invoice("out_invoice")
+        self.assertTrue(invoice.ubl_cii_xml_id)
+        self.assertEqual(invoice.ubl_cii_xml_id.name[-11:], "cius_ro.xml")
+        self._assert_invoice_attachment(invoice.ubl_cii_xml_id, xpaths=None, expected_file_path='from_odoo/ciusro_out_invoice.xml')
+
+    def test_export_constraints(self):
+        def create_basic_invoice():
```
Comment (malb-odoo) [inline addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:129]: does this function still needed ?

**Review by malb-odoo — CHANGES_REQUESTED** [review]
Thanks for the changes ! :smile: Few more remarks and i think it will be good to go

**Review by malb-odoo — APPROVED** [review]
LGTM nice work ! :smile:

File: addons/l10n_account_edi_ubl_cii_tests/tests/test_xml_cii_fr.py [hunk 5]
```diff
@@ -182,7 +182,7 @@ def test_export_import_invoice(self):
                         <PaymentReference>___ignore___</PaymentReference>
                 </xpath>
             ''',
-            expected_file='from_odoo/facturx_out_invoice.xml',
+            expected_file_path='from_odoo/facturx_out_invoice.xml',
```
Comment (smetl) [inline addons/l10n_account_edi_ubl_cii_tests/tests/test_xml_cii_fr.py:185]: remove the keywords instead. It's not an optional parameter

File: addons/l10n_ro_edi/data/cius_ro_templates.xml [hunk 6]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="utf-8"?>
+<odoo>
+
+    <template id="cius_ro_InvoiceType" inherit_id="account_edi_ubl_cii.ubl_21_InvoiceType" primary="True">
+        <xpath expr="//*[local-name()='DocumentCurrencyCode']" position="after">
+            <cbc:TaxCurrencyCode
+                    xmlns:cbc="urn:oasis:names:specification:ubl:schema:xsd:CommonBasicComponents-2"
+                    xmlns="urn:oasis:names:specification:ubl:schema:xsd:Invoice-2"
+                    t-out="vals.get('tax_currency_code')"/>
```
Comment (smetl) [inline addons/l10n_ro_edi/data/cius_ro_templates.xml:9]: Should be added in ubl_cii version 2.0 because part of the standard template:
http://www.datypic.com/sc/ubl20/t-ns19_InvoiceType.html

File: addons/l10n_ro_edi/i18n/l10n_ro_edi.pot [hunk 7]
```diff
@@ -0,0 +1,56 @@
+# Translation of Odoo Server.
```
Comment (smetl) [inline addons/l10n_ro_edi/i18n/l10n_ro_edi.pot:1]: useless in master, to remove

File: addons/l10n_ro_edi/i18n/ro.po [hunk 8]
```diff
@@ -0,0 +1,60 @@
+# Translation of Odoo Server.
```
Comment (smetl) [inline addons/l10n_ro_edi/i18n/ro.po:1]: useless in master, to remove

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+
+    def _export_invoice_filename(self, invoice):
+        return f"{invoice.name.replace('/', '_')}_cius_ro.xml"
+
+    def _get_partner_address_vals(self, partner):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals = super()._get_partner_address_vals(partner)
+
+        if partner.state_id:
+            vals["country_subentity"] = partner.country_code + '-' + partner.state_id.code
```
Comment (smetl) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:22]: `'country_subentity': partner.state_id.name,`
I really doubt this is necessary to put a custo for that. Did you check regarding a proper validator like ecosio?

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    def _get_invoice_tax_totals_vals_list(self, invoice, taxes_vals):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals_list = super()._get_invoice_tax_totals_vals_list(invoice, taxes_vals)
+
+        if invoice.currency_id.name != 'RON':
+            # add new TaxTotal in RON
+            from_currency = invoice.currency_id
+            ron_currency = self.env.ref("base.RON")
+            from_amount = vals_list[0]['tax_amount']
+            to_amount = from_currency._convert(from_amount, ron_currency, invoice.company_id, invoice.date, False)
```
Comment (smetl) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:35]: Noooo! Everything you need is already in `taxes_vals` in both currencies. 
Take a look at `_get_invoice_tax_totals_vals_list` in 2.0 template. Just remove the trailing _currency and you should get the correct result

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+            return [{
+                'company_id': company_registry,
+                'TaxScheme_vals': {},
+                'tax_scheme_id': 'VAT',
+            }]
+
+        if partner.vat and not partner.vat.startswith(partner.country_code):
+            for vals in vals_list:
+                vals['company_id'] = partner.country_code + partner.vat
```
Comment (smetl) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:62]: Incorrect. You never know how partner.vat is formatted. Stick to the already computed values in the xml...
Most of this code is probably standard somehow.

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+    def _get_partner_party_legal_entity_vals_list(self, partner):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals_list = super()._get_partner_party_legal_entity_vals_list(partner)
+
+        for vals in vals_list:
+            if not partner.vat and partner.company_registry:
+                vals['company_id'] = partner.company_registry
+            if vals['company_id'] and not vals['company_id'].startswith(partner.country_code):
+                vals['company_id'] = partner.country_code + vals['company_id']
```
Comment (smetl) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:74]: same here...

File: addons/l10n_ro_edi/i18n/l10n_ro_edi.pot (hunk 7 above)
Comment (malb-odoo) [reply addons/l10n_ro_edi/i18n/l10n_ro_edi.pot:1]: Not for localization, also he adds constraint i think it's worth to have a po

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py (hunk 9 above)
Comment (yosuanicolaus) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:22]: that's one of the rule from the schematron. it has to be in that format.
image

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+        return vals_list
+
+    def _export_invoice_vals(self, invoice):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals = super()._export_invoice_vals(invoice)
+
+        vals['InvoiceType_template'] = 'l10n_ro_edi.cius_ro_InvoiceType'
+        vals['vals']['customization_id'] = 'urn:cen.eu:en16931:2017#compliant#urn:efactura.mfinante.ro:CIUS-RO:1.0.1'
+        vals['vals']['tax_currency_code'] = 'RON'
```
Comment (JulienVR) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:84]:
```suggestion
        vals['vals'].update({
           'customization_id': 'urn:cen.eu:en16931:2017#compliant#urn:efactura.mfinante.ro:CIUS-RO:1.0.1',
           'tax_currency_code': 'RON',
        })
```

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py (hunk 10 above)
Comment (JulienVR) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:35]: I'm not sure about this: if you have an invoice which is not in RON currency (let's say it's in EUR), you will have every amounts in the xml in EUR except the TaxAmount in the TaxTotals ? Is it what is expected ?

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        return vals_list
+
+    def _get_partner_party_tax_scheme_vals_list(self, partner, role):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals_list = super()._get_partner_party_tax_scheme_vals_list(partner, role)
+
+        if not partner.vat and partner.company_registry:
+            company_registry = partner.company_registry
+            if not company_registry.startswith(partner.country_code):
+                company_registry = partner.country_code + company_registry
```
Comment (JulienVR) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:52]: You could add some constraints on the company_registry in `_export_invoice_constraints` to avoid doing this check :)
FYI, we didn't do it in `account_edi_ubl_cii` because before saas-16.2, the constraints weren't blocking.

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py (hunk 11 above)
Comment (JulienVR) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:62]: ... or add a constraint as I suggested above

**Review by JulienVR — COMMENTED** [review]
Good job ! A few details but I can see you got the idea of how `account_edi_ubl_cii` works :)

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py (hunk 10 above)
Comment (yosuanicolaus) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:35]: @JulienVR We will still have the normal TaxTotal in EUR but add another TaxTotal in RON, so in total there will be 2 TaxTotal. [Example here](https://github.com/odoo/odoo/pull/130617/files/26b1e453633c889c8312a62ddbe510de5a431ef1#diff-dcc450ce3734bd538eb8f1fdf70003f4dd27c2fde910b614d4f19dcc1673c4b1R107-R123)

File: addons/l10n_account_edi_ubl_cii_tests/tests/test_xml_cii_fr.py (hunk 5 above)
Comment (yosuanicolaus) [reply addons/l10n_account_edi_ubl_cii_tests/tests/test_xml_cii_fr.py:185]: this one is cherry picked from Brice's commit in the OIOUBL PR. I will re-cherry pick it once he changes it

File: addons/l10n_ro_edi/i18n/l10n_ro_edi.pot (hunk 7 above)
Comment (malb-odoo) [reply addons/l10n_ro_edi/i18n/l10n_ro_edi.pot:1]: For me it needs to be there

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+
+    def _get_partner_party_tax_scheme_vals_list(self, partner, role):
+        # EXTENDS account.edi.xml.ubl_bis3
+        vals_list = super()._get_partner_party_tax_scheme_vals_list(partner, role)
+
+        if not partner.vat and partner.company_registry:
+            company_registry = partner.company_registry
+
+            return [{
+                'company_id': company_registry,
```
Comment (JulienVR) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:48]: :thinking:

File: addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+                f"ciusro_{partner_type}_city_required": self._check_required_fields(partner, 'city'),
+                f"ciusro_{partner_type}_street_required": self._check_required_fields(partner, 'street'),
+                f"ciusro_{partner_type}_state_id_required": self._check_required_fields(partner, 'state_id'),
+            })
+
+            if not partner.vat and not partner.company_registry:
+                constraints[f"ciusro_{partner_type}_tax_identifier_required"] = _(
+                    "The following partner doesn't have a VAT nor Company ID: %s. "
+                    "At least one of them is required. "
+                ) % partner.name
```
Comment (JulienVR) [inline addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:93]: So we no longer verify that the company_registry starts with the country_code ? I thought it was required ?
Same for the VAT.

Comment (yosuanicolaus) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:91]: @JulienVR  sorry about this :laughing: at first I thought this was actually not needed but I rechecked again and turns out the rule actually exist. I added it on the newest commit :)

Comment (malb-odoo) [reply addons/l10n_ro_edi/models/account_edi_xml_ubl_ciusro.py:91]: @yosuanicolaus Don't forget to squash it when the pr is ready :smile: And for next time instead of making a new commit for reviewer to check you can check how git amend works :smile:

**Comment by yosuanicolaus** [comment]
@smetl @JulienVR ready for review :)

File: addons/l10n_account_edi_ubl_cii_tests/tests/common.py [hunk 17]
```diff
@@ -225,7 +225,7 @@ def _generate_move(self, seller, buyer, send=True, **invoice_kwargs):
             account_move._generate_pdf_and_send_invoice(self.move_template)
         return account_move
 
-    def _assert_invoice_attachment(self, attachment, xpaths, expected_file):
+    def _assert_invoice_attachment(self, attachment, xpaths, expected_file_path):
```
Comment (JulienVR) [inline addons/l10n_account_edi_ubl_cii_tests/tests/common.py:228]: either cherry-pick Brice's commit or make a separate commit for this changes, but don't squash it with your other changes, it has nothing to do with the Romanian EDI :)

**Review by JulienVR — APPROVED** [review]
Last thing and it's good for me :+1:, thank you !

**Review by antoine162 — COMMENTED** [review]
Just one thing: the `TestUBLRO.test_export_constraints` test seems to be failing at the moment :smile:

The Community runbot doesn't install modules starting with 'l10n' and because this is a Community PR, the 'l10n' build isn't launched by default, so it's quite easy to miss broken tests.

```log
2023-08-31 10:58:59,625 428549 ERROR test-ro-master odoo.addons.l10n_ro_edi.tests.test_xml_ubl_ro: FAIL: TestUBLRO.test_export_constraints
odoo.exceptions.UserError: Errors occured while creating the EDI document (format: CIUS RO):
The following partner's doesn't have a country code prefix in their Company ID: Hudson Construction. 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/odoo/src/multiverse/master/odoo/addons/l10n_ro_edi/tests/test_xml_ubl_ro.py", line 150, in test_export_constraints
    invoice._generate_pdf_and_send_invoice(self.move_template, from_cron=False, allow_fallback_pdf=False)
AssertionError: "required" does not match "Errors occured while creating the EDI document (format: CIUS RO):
The following partner's doesn't have a country code prefix in their Company ID: Hudson Construction. "
```

**Review by smetl — APPROVED** [review]
tiny typo, otherwise, it's ok

File: addons/l10n_ro_edi/tests/test_xml_ubl_ro.py [hunk 18]
```diff
@@ -0,0 +1,147 @@
+from odoo import Command
+from odoo.addons.l10n_account_edi_ubl_cii_tests.tests.common import TestUBLCommon
+from odoo.addons.account.tests.test_account_move_send import TestAccountMoveSendCommon
+from odoo.exceptions import UserError
+from odoo.tests import tagged
+
+
+@tagged('post_install_l10n', 'post_install', '-at_install')
+class TestUBLRO(TestUBLCommon, TestAccountMoveSendCommon):
```
Comment (smetl) [inline addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:9]: why `TestAccountMoveSendCommon` ?

File: addons/l10n_ro_edi/tests/test_xml_ubl_ro.py [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+            'amount': 19,
+            'type_tax_use': 'sale',
+            'country_id': cls.env.ref('base.ro').id,
+        })
+
+    ####################################################
+    # Test export - import
+    ####################################################
+
+    def create_move(self, move_type):
```
Comment (smetl) [inline addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:57]: use `_generate_move` instead?

Comment (yosuanicolaus) [reply addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:57]: @smetl
I think that would result in more code since I still need to add a few lines for the parameters. I prefer the current way as it's simpler :)
Or do you mean using _generate_move inside the create_move?

Comment (smetl) [reply addons/l10n_ro_edi/tests/test_xml_ubl_ro.py:56]:
```py
def create_move(self, move_type):
    return self._generate_move(
        self.env.company.partner_id,
        self.partner_a,
        send=True,
        move_type=move_type,
        invoice_line_ids=[
            Command.create({
                'product_id': self.product_a.id,
                'price_unit': 500.0,
                'tax_ids': [Command.set(self.tax_19.ids)],
            }),
            Command.create({
                'product_id': self.product_b.id,
                'price_unit': 1000.0,
                'tax_ids': [Command.set(self.tax_19.ids)],
            }),
        ],
    )
```

**Review by smetl — APPROVED** [review]
@robodoo r+

**Comment by smetl** [comment]
@robodoo rebase-ff
